### PR TITLE
RCHAIN-3250 extract Engine

### PR DESCRIPTION
**Pull request by rabbitonweb**
## Overview
Extracts concept of an Engine that traverse through states during casper lifecycle. Eventually `MultiparentCasperRef` will get obsolete and will become part of Engine as well.

For now `Engine` represents transitions that casper does during its lifecycle. Different engines behave differently to incoming message. Final state is the state called `Running`.

Once this PR gets in, RCHAIN-3250 can be fixed as it will be a matter of introduce intermediate step in transition between `GenesisValidator` and `Running`

### JIRA ticket:
RCHAIN-3250

### CODE REVIEW

**Comment by rabbitonweb** [comment]
This is now ready for review

**Comment by rabbitonweb** [comment]
bors try

File: casper/src/main/scala/coop/rchain/casper/engine/CasperLaunch.scala [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+        val action = connectToExistingNetwork[F](approvedBlock, init)
+        (msg, action)
+      case None if (init.conf.approveGenesis) =>
+        val msg    = "ApprovedBlock not found in storage, taking part in ceremony as genesis validator"
+        val action = connectAsGenesisValidator[F](init)
+        (msg, action)
+      case None if (init.conf.createGenesis) =>
+        val msg =
+          "ApprovedBlock not found in storage, taking part in ceremony as ceremony master"
+        val action = initBootstrap[F](init, toTask)
```
Comment (KentShikama) [inline casper/src/main/scala/coop/rchain/casper/engine/CasperLaunch.scala]: Perhaps rename to initCeremonyMaster

File: casper/src/main/scala/coop/rchain/casper/engine/Running.scala [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+  }
+
+  private def handleBlockMessage(peer: PeerNode, b: BlockMessage): F[Unit] =
+    for {
+      isOldBlock <- MultiParentCasper[F].contains(b)
+      _ <- if (isOldBlock)
+            Log[F].info(s"Received block ${PrettyPrinter.buildString(b.blockHash)} again.")
+          else
+            handleNewBlock(peer, b)
+
```
Comment (KentShikama) [inline casper/src/main/scala/coop/rchain/casper/engine/Running.scala]: Unnecessary new line

File: docs/features.md [hunk 3]
```diff
@@ -137,7 +137,7 @@
 * `validatorC` joins p2p, pointing on `ceremonyMaster` as bootstrap
 * `validatorC` sends `ApprovedBlockRequest` to `ceremonyMaster`
 * `ceremonyMaster` sends `ApprovedBlock` to `validatorC`
-* `validatorC` transitions to `ApprovedBlockReceivedHandler`
+* `validatorC` transitions to `Running`
```
Comment (KentShikama) [inline docs/features.md]: Glad you remembered to update features.md

File: casper/src/main/scala/coop/rchain/casper/engine/CasperLaunch.scala (hunk 1 above)
Comment (rabbitonweb) [reply casper/src/main/scala/coop/rchain/casper/engine/CasperLaunch.scala]: next pr

**Pull request merged**